### PR TITLE
Suspend Solution Crawler during Find All References

### PR DESCRIPTION
**Pull request by rchande**
In #2876, VS is unresponsive and constantly GCing. FAR and Solution
Crawler are both running and serializing trees (from different
projects), which is allocating and causing the GCs. To help mitigate
this, the FAR command handler will start a global operation before
starting the search.

Can you take a look? @pharring @heejaechang 
cc: @Pilchie @jasonmalinowski @dpoeschl @brettfo @balajikris

### CODE REVIEW

**Comment by rchande** [comment]
I wasn't sure if we should just make the FAR engine itself start a global operation. Thoughts?

**Comment by jasonmalinowski** [comment]
The engine itself wouldn't really be the place to do it I think. Otherwise highlight references would keep restarting global operations all the time as well.

**Comment by rchande** [comment]
Yeah, that would not be good.

**Comment by heejaechang** [comment]
:+1: 

by the way, solution crawler already pause if wait dialog shows up. so. this must be when FAR running but wait dialog is not up yet.

**Comment by jasonmalinowski** [comment]
Per #2876 a wait dialog already was up....so is something else up?

**Comment by rchande** [comment]
@tmat says the dialog hadn't yet rendered. Maybe it hadn't signalled solution crawler yet? @heejaechang how does it get notified?

**Comment by heejaechang** [comment]
I don’t think Find All Reference Command Handler uses wait dialog box.

File: src/EditorFeatures/Core/Implementation/FindReferences/FindReferencesCommandHandler.cs [hunk 1]
```diff
@@ -51,12 +52,16 @@ internal void FindReferences(ITextSnapshot snapshot, int caretPosition)
                     var service = document.Project.LanguageServices.GetService<IFindReferencesService>();
                     if (service != null)
                     {
-                        if (!service.TryFindReferences(document, caretPosition, context))
+                        var globalOperationService = document.Project.Solution.Workspace.Services.GetService<IGlobalOperationNotificationService>();
+                        using (globalOperationService?.Start("FindReferences"))
```
Comment (jaredpar) [inline src/EditorFeatures/Core/Implementation/FindReferences/FindReferencesCommandHandler.cs]: Does this string value matte?  i.e. is the code on the other end looking for this exact value?  If so can we make it a constant?

**Comment by rchande** [comment]
It uses a wait dialog (from the same file):
```
 _waitIndicator.Wait(
                title: EditorFeaturesResources.FindReferences,
                message: EditorFeaturesResources.FindingReferences,
                action: context =>
            {
                Document document = snapshot.GetOpenDocumentInCurrentContextWithChanges();
                if (document != null)
                {
                    var service = document.Project.LanguageServices.GetService<IFindReferencesService>();
                    if (service != null)
                    {
                        var globalOperationService = document.Project.Solution.Workspace.Services.GetService<IGlobalOperationNotificationService>();
                        using (globalOperationService?.Start("FindReferences"))```

**Comment by heejaechang** [comment]
yep. just saw that. diff was hiding it. :) so, I think the dialog box wasnt rendered yet.

**Comment by jasonmalinowski** [comment]
Why does the dialog being rendered matter? How does the solution crawler currently detect that the dialog is up?

File: src/EditorFeatures/Core/Implementation/FindReferences/FindReferencesCommandHandler.cs (hunk 1 above)
Comment (rchande) [inline src/EditorFeatures/Core/Implementation/FindReferences/FindReferencesCommandHandler.cs]: @jaredpar The value doesn't matter, but I will switch to using ```EditorFeaturesResources.FindReferences```.

**Comment by rchande** [comment]
Supposedly VisualStudioWaitContext starts an operation as soon as it's constructed, but solution crawler is clearly running in that dump.

**Comment by rchande** [comment]
http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/Utilities/VisualStudioWaitContext.cs,37

**Comment by heejaechang** [comment]
actually not sure. just checked. 

http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/Utilities/VisualStudioWaitContext.cs,a0f752ca3262764b

it actually doesn't care. but the notification is async and solution crawler is cancellation based, so possible while solution crawler is already told paused, but not cancelled yet. and de-serialization happen on compilation creation which is not incremental.

**Comment by rchande** [comment]
The deserialization is also not cancellable, so it's possible that solution crawler would cancel but still wind up spending a long time deserializing before it could actually check the cancellation token. Maybe we should be plumbing cancellation tokens through the deserialize code?

File: src/EditorFeatures/Core/Implementation/FindReferences/FindReferencesCommandHandler.cs (hunk 1 above)
Comment (heejaechang) [inline src/EditorFeatures/Core/Implementation/FindReferences/FindReferencesCommandHandler.cs]: the string will pass down to listener. currently there is no one care though. probably should log that information to etl.

Comment (pharring) [inline src/EditorFeatures/Core/Implementation/FindReferences/FindReferencesCommandHandler.cs]: I'd leave it the way it is, Ravi. We don't usually localize such strings. As HeeJae says, they're useful for logging, but never displayed to the user.

**Comment by Pilchie** [comment]
@heejaechang Could the serialization seen in this trace be the result of the cache race condition you just hit?

At the moment, I can't advise taking this, because we don't have any reason to believe it will have any effect beyond what the Wait dialog already provides.

**Comment by heejaechang** [comment]
Oh… possibly!


-          heejae

**Comment by rchande** [comment]
@heejaechang Are you looking at that race condition? If so, I'm going to resolve @tmat's bug and close this PR.

**Comment by heejaechang** [comment]
It is in master.


-          heejae

**Comment by rchande** [comment]
You mean it's fixed?

**Comment by heejaechang** [comment]
Yes

Sent from my Windows Phone
________________________________

**Comment by Pilchie** [comment]
Master or stabilization?

**Comment by heejaechang** [comment]
Master.

Sent from my Windows Phone
________________________________

**Comment by heejaechang** [comment]
It just got green light on stabilization but didn't merge yet

Sent from my Windows Phone
________________________________
From: HeeJae Chang<mailto:Heejae.Chang@microsoft.com>
Sent: ‎5/‎29/‎2015 5:15 PM
To: dotnet/roslyn<mailto:reply@reply.github.com>; dotnet/roslyn<mailto:roslyn@noreply.github.com>
Subject: RE: [roslyn] Suspend Solution Crawler during Find All References (#3153)

Master.

Sent from my Windows Phone
________________________________

**Comment by rchande** [comment]
Not the right solution, HeeJae has something pending.

**Pull request closed without merging**